[APO-1040] Prune edges when node fail during serialization

vincent0426 opened the pull request:
# Fix node invalidation logic to only mark invalid when serialized_node is null

## Summary

This PR refines the node invalidation logic during workflow serialization to be more granular. Previously, nodes were marked as invalid (and their edges pruned) for ANY error during serialization. Now, nodes are only marked invalid when the entire serialization completely fails (`serialized_node` is null), not for attribute-level validation errors.

**Key Changes:**
- **Modified error handling**: Separated attribute-level errors from complete node failures
- **Updated invalidation logic**: Only mark nodes invalid when `NotImplementedError` or `NodeValidationError` are raised during `serialize()`
- **Fixed test expectations**: Updated test to expect edges to remain when only attribute errors occur
- **Added comprehensive test**: New test verifies edges are pruned only for complete node failures

This addresses feedback from @dvargas92495 that edges shouldn't be removed for attribute-level errors like missing files, only when the entire node is lost.

## Review & Testing Checklist for Human

- [ ] **Verify behavioral change is correct**: Manually test that nodes with attribute errors (like missing files) retain their edges while nodes with complete serialization failures have edges pruned
- [ ] **Test existing workflows**: Ensure existing workflows that may have relied on the old invalidation behavior still work correctly  
- [ ] **Validate new test scenario**: Confirm the new test `test_serialize_node__with_complete_node_failure_prunes_edges` actually tests complete node serialization failure (not just attribute errors)
- [ ] **Edge case testing**: Test various error scenarios to ensure the distinction between attribute errors and complete failures works as expected

---

### Diagram

```mermaid
%%{ init : { "theme" : "default" }}%%
graph TD
    A["base_workflow_display.py<br/>serialize() method"]:::major-edit
    B["types.py<br/>add_error() method"]:::major-edit  
    C["code_execution_node.py<br/>error handling"]:::minor-edit
    D["test_code_execution_node.py<br/>test expectations"]:::major-edit
    E["Node Serialization Flow"]:::context
    
    A -->|"calls serialize() on nodes"| E
    E -->|"attribute error"| B
    E -->|"complete failure"| A
    A -->|"marks node invalid only<br/>for complete failures"| F["Invalid Node Tracking"]:::context
    B -->|"logs error without<br/>marking node invalid"| G["Error Collection"]:::context
    C -->|"reports missing file<br/>as attribute error"| B
    D -->|"tests both<br/>scenarios"| H["Test Coverage"]:::context

    subgraph Legend
        L1[Major Edit]:::major-edit
        L2[Minor Edit]:::minor-edit  
        L3[Context/No Edit]:::context
    end

    classDef major-edit fill:#90EE90
    classDef minor-edit fill:#87CEEB  
    classDef context fill:#FFFFFF
```

### Notes

- This change affects the fundamental behavior of workflow serialization error handling
- The distinction between attribute-level errors and complete node failures is critical for correct edge pruning
- All tests pass including the new test that specifically validates complete node failure scenarios
- Pre-commit hooks and CI checks all passed successfully

**Session Info**: Requested by vincent@vellum.ai (@vincent0426)  
**Devin Session**: https://app.devin.ai/sessions/b7412b8379bf45eab659d4de8325584c

vincent0426 committed:
Prune edges when node fail during serialization

vincent0426 committed:
Add is node invalid

dvargasfuertes commented on ee/vellum_ee/workflows/display/nodes/vellum/tests/test_code_execution_node.py:
```diff
@@ -253,16 +253,7 @@ class Workflow(BaseWorkflow):
-                    "type": "DEFAULT",
-                }
-            ],
+            "edges": [],
```
[blocking] we only need to do this if we lose the node entirely. now that our `add_error` is granularly scoped to just the attribute, the edge is now valid

devin-ai-integration[bot] committed:
Fix node invalidation logic to only mark invalid when serialized_node is null

- Separate attribute-level errors from node-level failures
- Only mark nodes invalid when entire serialization fails (NotImplementedError/NodeValidationError)
- Fix test to expect edges when only attribute errors occur
- Add test for complete node failure that verifies edges are pruned
- Addresses feedback from dvargas92495

Co-Authored-By: vincent@vellum.ai <0426vincent@gmail.com>

devin-ai-integration[bot] committed:
Move complete node failure test to workflow display tests

- Move test from test_code_execution_node.py to test_workflow_display.py
- Use BaseNode instead of CodeExecutionNode for simpler test approach
- Fix mypy errors by importing JsonObject from correct module
- Test verifies edges are pruned when node completely fails to serialize
- Addresses feedback to use simpler BaseNode approach in workflow tests

Co-Authored-By: vincent@vellum.ai <0426vincent@gmail.com>
